lib/aws/AwsSigv4Signer.js

```diff
@@ -10,33 +10,80 @@
  */
 
 'use strict';
-const Connection = require('../Connection');
 const aws4 = require('aws4');
+const Transport = require('../Transport');
+const Connection = require('../Connection');
 const AwsSigv4SignerError = require('./errors');
 
-function AwsSigv4Signer(opts) {
-  if (opts && (!opts.region || opts.region === null || opts.region === '')) {
+function AwsSigv4Signer(awsConfig) {
+  if (
+    (!awsConfig.region || awsConfig.region === null || awsConfig.region === '') &&
+    typeof awsConfig.getCredentials !== 'function'
+  ) {
     throw new AwsSigv4SignerError('Region cannot be empty');
   }
-  if (opts && (!opts.credentials || opts.credentials === null || opts.credentials === '')) {
+  if (
+    (!awsConfig.credentials || awsConfig.credentials === null || awsConfig.credentials === '') &&
+    typeof awsConfig.getCredentials !== 'function'
+  ) {
     throw new AwsSigv4SignerError('Credentials cannot be empty');
   }
 
   function buildSignedRequestObject(request = {}) {
     request.service = 'es';
-    request.region = opts.region;
+    request.region = awsConfig.region;
     request.headers = request.headers || {};
     request.headers['host'] = request.hostname;
-    return aws4.sign(request, opts.credentials);
+    return aws4.sign(request, awsConfig.credentials);
   }
   class AwsSigv4SignerConnection extends Connection {
     buildRequestObject(params) {
       const request = super.buildRequestObject(params);
       return buildSignedRequestObject(request);
     }
   }
+
+  function awaitAwsCredentials(awsConfig) {
+    return new Promise((resolve, reject) => {
+      awsConfig.getCredentials((err) => {
+        err ? reject(err) : resolve();
+      });
+    });
+  }
+
+  class AwsSigv4SignerTransport extends Transport {
+    request(params, options = {}, callback = undefined) {
+      // options is optional,
+      // so if it is omitted, options will be the callback
+      if (typeof options === 'function') {
+        callback = options;
+        options = {};
+      }
+
+      // Check if getCredentials exists,
+      // if so this is an aws-sdk v2 global config object
+      if (typeof awsConfig.getCredentials !== 'function') {
+        if (typeof callback === 'undefined') {
+          return super.request(params, options);
```

Comment:
Maybe I'm missing something here, but shouldn't we be refreshing credentials in the V3 version like we do in the V2 version? How is this going to prevent the expired credentials case?

Reply:
You are right, I'm not using aws v3 and this is taken from aws-opensearch-connector library but I had the same thought when checking their code originally.. and since I don't know I assume I was missing something that changed with aws v3.
I'm not sure, we would need the help from someone who knows better.

Reply:
ok, so a bit of research on aws v3 I found the reason why STS and assumeRole are given in example for v3
AWS handle refreshing the token behind the scene using [ChainableTemporaryCredentials](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/ChainableTemporaryCredentials.html)
Maybe we should mention this on the example for aws v3 to makes sure consumer are aware of this.